testing/conformance/conformance/applications.py

```diff
@@ -50,17 +77,644 @@ class WindowDetectorPython3(Application):
+    def parse_output(self, bs):
+        logging.debug('parse_output: {}'.format(repr(bs)))
+        return json.loads(bs[4:].decode())['value']
+
+    def collect(self):
+        data = super().collect(None) # {'0': [...]}
+        output = [self.parse_output(v) for v in data[0]]
+
+        return output
+
+################################
+# Resilience Test Applications #
+################################
+try:
+    basestring
+except:
+    basestring = str
+
+
+def get_await_values(last_sent_groups):
+    logging.debug(repr(last_sent_groups))
+    return [(key,
+             "[{}]".format(",".join(
+                 ("{}".format(max(0,x)) for x in range(val-3, val+1)))))
+            for group in last_sent_groups
+            for key, val in group]
+
+
+# TODO: refactor and move to control.py
+def pause_senders_and_sink_await(cluster, timeout=10):
+    logging.debug("pause_senders_and_sink_await")
+    if not cluster.senders:
+        logging.debug("No senders to pause. Continuing.")
+        return
+    # we currently only support one type of sender per run
+    if not isinstance(cluster.senders[0], Sender):
+        return
+    cluster.pause_senders()
+    for s in cluster.senders:
+            logging.debug("Sender paused with {}, and {} bytes in buffer"
+                .format(s.reader.gen.seqs, len(s.batch)))
+    logging.debug("Waiting for messages to propagate to sink")
+    await_values = get_await_values([sender.last_sent() for sender in cluster.senders])
+    cluster.sink_await(values=await_values, func=json_keyval_extract)
+    # Since snapshots happen at a 1 second frequency, we need to wait
+    # more than 1 second to guarantee a snapshot after messages arrived
+    logging.debug("All messages arrived at sink!")
+
+
+class ResilienceOperation(object):
+    """
+    Baseclass for resilience operation types
+    """
+    def __init__(self, size, check_size=True):
+        if check_size:
+            if isinstance(size, int):
+                if size <= 0:
+                    raise ValueError("size must be a positive integer")
+            else:
+                raise TypeError("size must be an integer")
+        self.size = size
+
+    def sign(self):
+        """
+        Identify the sign of the operation: e.g. whether it adds or subtracts
+        from the total size of the cluster
+        >0: 1
+        0: 0
+        <1: -1
+        """
+        raise NotImplementedError
+
+    def apply(self, cluster, data=None):
+        """
+        The logic involved in applying this step and any direct validations
+        """
+        raise NotImplementedError
+
+    def name(self):
+        """
+        The name of of the operation, along with its size.
+        Useful when printing a history of operations in a concise manner.
+        """
+        return "{}:{}".format(self.__class__.__name__, self.size)
+
+    def __str__(self):
+        return self.name()
+
+
+class Grow(ResilienceOperation):
```

Comment:
Can you move these operations to a separate file?

Reply:
I broke them down to individual apps.
I still left the operations together with resilience testing app file, but they're entirely separate from the cli, topology, and window testing apps and their associated files.
It could be broken down further into submodules if necessary.